websocket.go

```diff
@@ -436,6 +456,155 @@ func (c *Client) SendProtobuf(dst *network.ServerIdentity, msg interface{}, ret
+	// StartNode will be applied before shuffling.
+	//   Default: false
+	DontShuffle bool
+}
+
+// GetList returns how many requests to start in parallel and a channel of nodes to be used.
+// If po == nil, it uses default values.
+func (po *ParallelOptions) GetList(nodes []*network.ServerIdentity) (parallel int, nodesChan chan *network.ServerIdentity) {
+	// Default values
+	parallel = (len(nodes) + 1) / 2
+	askNodes := len(nodes)
+	startNode := 0
+	var ignoreNodes []*network.ServerIdentity
+	var perm []int
+	if po != nil {
+		if po.Parallel > 0 && po.Parallel < parallel {
+			parallel = po.Parallel
+		}
+		if po.StartNode > 0 && po.StartNode < len(nodes) {
+			startNode = po.StartNode
+			askNodes -= startNode
+		}
+		if po.AskNodes > 0 && po.AskNodes < len(nodes) {
+			askNodes = po.AskNodes
+		}
+		if askNodes < parallel {
+			parallel = askNodes
+		}
+		if po.DontShuffle {
+			for i := range nodes {
+				perm = append(perm, i)
+			}
+		}
+		ignoreNodes = po.IgnoreNodes
+	}
+	if len(perm) == 0 {
+		perm = rand.Perm(len(nodes))
+	}
+
+	nodesChan = make(chan *network.ServerIdentity, askNodes)
+	for i := range nodes {
+		addNode := true
+		node := nodes[(startNode+perm[i])%len(nodes)]
+		for _, ignore := range ignoreNodes {
+			if node.Equal(ignore) {
+				addNode = false
+				break
+			}
+		}
+		if addNode {
+			nodesChan <- node
+		}
+		if len(nodesChan) == askNodes {
+			break
+		}
+	}
+	return parallel, nodesChan
+}
+
+// Quit return false if po == nil, or the value in po.QuitError.
+func (po *ParallelOptions) Quit() bool {
+	if po == nil {
+		return false
+	}
+	return po.QuitError
+}
+
+// SendProtobufParallel sends the msg to a set of nodes in parallel and returns the first successful
+// answer. If all nodes return an error, only the first error is returned.
+// The behaviour of this method can be changed using the ParallelOptions argument. It is kept
+// as a structure for future enhancements. If opt is nil, then standard values will be taken.
+func (c *Client) SendProtobufParallel(nodes []*network.ServerIdentity, msg interface{}, ret interface{},
+	opt *ParallelOptions) (*network.ServerIdentity, error) {
+	buf, err := protobuf.Encode(msg)
+	if err != nil {
+		return nil, err
+	}
+	path := strings.Split(reflect.TypeOf(msg).String(), ".")[1]
+
+	parallel, nodesChan := opt.GetList(nodes)
+	errChan := make(chan error, parallel)
+	replyChan := make(chan []byte, parallel)
+	siChan := make(chan *network.ServerIdentity, parallel)
+	finish := &sync.Once{}
+	for g := 0; g < parallel; g++ {
+		go func() {
+			for {
+				var node *network.ServerIdentity
+				select {
+				case node = <-nodesChan:
+				default:
+				}
+				if node == nil {
+					return
+				}
+				log.Lvlf2("Asking %T from: %v - %v", msg, node.Address, node.URL)
+				reply, err := c.Send(node, path, buf)
+				if err != nil {
+					log.Lvl2("Error while sending to node:", node, err)
+					errChan <- err
+				} else {
+					finish.Do(func() { close(nodesChan) })
```

Comment:
If all the requests fail, the go routines will be stuck trying to read the next node (as there is no node left)

Reply:
Are you referring to the lines
```
case node = <-nodesChan:
default:
```
?
Then it's OK, because the `default` is the fall-through option. I can put the `return` in the default. But it seemed to me that a `nil` node can appear in the channel... Perhaps not.

Reply:
Now I know why the `if node == nil` came after the select:
- if there is a value, `node` is populated
- if the channel is empty, it goes in `default`, and `node == nil`
- if the channel is closed, `node` is populated with `nil` !?!